### PR TITLE
Implement elevator states

### PR DESCRIPTION
**Pull request by bananapizzza**

### CODE REVIEW

File: src/main/java/ElevatorManagementSystem.java [hunk 1]
```diff
@@ -3,19 +3,28 @@
...
 public class ElevatorManagementSystem {
     @NonNull
     int numOfFloor;
+    @NonNull
+    int numOfBasementFloor;
+    @NonNull
+    int numOfElevators;
 
-    final List<Elevator> elevatorList = new ArrayList<>();
+    public final int DIRECTION_GOING_UP = 1;
```
Comment (xaveng) [inline src/main/java/ElevatorManagementSystem.java]: Minor: I would create an Enum Direction with these two values.

Comment (xaveng) [reply src/main/java/ElevatorManagementSystem.java]: You've done it in the next commit. Nice!

File: src/main/java/MovingState.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+public class MovingState implements ElevatorState {
+    @Override
+    public boolean selectFloor(Elevator elevator, int floor) {
+        if (!elevator.getPossibleFloors().contains(floor)) {
```
Comment (xaveng) [inline src/main/java/MovingState.java]: This test is been used multiple places. Can we pull this out or pull this up?

File: src/main/java/MovingState.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        } else {
+            System.out.println("Cannot add "+floor+" floor to the destination.");
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean openDoor() {
+        System.out.println("Cannot open the door right now.");
```
Comment (xaveng) [inline src/main/java/MovingState.java]: Nice!

File: src/main/java/MovingState.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        return false;
+    }
+
+    @Override
+    public boolean closeDoor() {
+        System.out.println("Cannot open the door right now.");
```
Comment (xaveng) [inline src/main/java/MovingState.java]: open -> close

File: src/main/java/Elevator.java [hunk 5]
```diff
@@ -1,24 +1,119 @@
...
+                System.out.println("Going up...");
+                this.direction = Direction.GOING_UP;
+            } else if (destinationFloor < currentFloor) {
+                System.out.println("Going down...");
+                this.direction = Direction.GOING_DOWN;
+            }
+            this.currentState = movingState;
+            if(this.direction.equals(Direction.GOING_UP)){
+                while(currentFloor < destinationFloor){
+                    try {
```
Comment (xaveng) [inline src/main/java/Elevator.java]: Let's pull this sleep feature to a function as well!

File: src/main/java/ElevatorManagementSystem.java (hunk 1 above)
Comment (bananapizzza) [reply src/main/java/ElevatorManagementSystem.java]: :D

File: src/main/java/MovingState.java (hunk 2 above)
Comment (bananapizzza) [reply src/main/java/MovingState.java]: in the last commit, I've changed it to 
if (!elevator.isTheFloorPossibleToGo(floor))

File: src/main/java/MovingState.java (hunk 3 above)
Comment (bananapizzza) [reply src/main/java/MovingState.java]: :D

File: src/main/java/MovingState.java (hunk 4 above)
Comment (bananapizzza) [reply src/main/java/MovingState.java]: Oh my mistake

File: src/main/java/Elevator.java (hunk 5 above)
Comment (bananapizzza) [reply src/main/java/Elevator.java]: Okie

Comment (bananapizzza) [reply src/main/java/Elevator.java]: Okie

File: src/test/java/ElevatorTest.java [hunk 6]
```diff
@@ -26,49 +28,62 @@ void test_selectFloor_IdleCase() {
...
 
     @Test
-    void test_selectFloor_MovingState() {
+    void test_selectFloor_MovingState() throws InterruptedException {
         elevator.setCurrentState(elevator.getMovingState());
         elevator.setDirection(Direction.GOING_UP);
+        elevator.start();
+
+        assertTrue(elevator.selectFloor(3));
+        Thread.sleep(3000);
```
Comment (xaveng) [inline src/test/java/ElevatorTest.java]: Can we use sleep_time*floor?

File: src/main/java/Elevator.java [hunk 7]
```diff
@@ -9,21 +8,29 @@
 @Getter
 @Setter
 @FieldDefaults(level = AccessLevel.PRIVATE)
-public class Elevator {
+public class Elevator extends Thread {
```
[the remark this answers is not in the document]
Comment (bananapizzza) [reply src/main/java/Elevator.java]: Changed.

File: src/main/java/Elevator.java [hunk 8]
```diff
@@ -57,53 +64,59 @@ boolean callEmergency() {
...
+                        setDirection(Direction.GOING_DOWN);
                     }
-                    currentFloor--;
-                    System.out.println("Current floor is " + currentFloor);
-                    if (destinationFloors[getIndex(currentFloor)]) {
-                        break;
+
+                    if (this.direction.equals(Direction.GOING_UP)) {
+                        while (currentFloor < destinationFloor) {
+                            waiting(MOVING_TIME);
```
[the remark this answers is not in the document]
Comment (bananapizzza) [reply src/main/java/Elevator.java]: Changed.

File: src/main/java/Elevator.java [hunk 9]
```diff
@@ -57,53 +64,59 @@ boolean callEmergency() {
...
+                        setDirection(Direction.GOING_UP);
+                    } else if (destinationFloor < currentFloor) {
+                        setDirection(Direction.GOING_DOWN);
                     }
-                    currentFloor--;
-                    System.out.println("Current floor is " + currentFloor);
-                    if (destinationFloors[getIndex(currentFloor)]) {
-                        break;
+
+                    if (this.direction.equals(Direction.GOING_UP)) {
```
[the remark this answers is not in the document]
Comment (bananapizzza) [reply src/main/java/Elevator.java]: Changed.

File: src/main/java/Elevator.java [hunk 10]
```diff
@@ -57,53 +64,59 @@ boolean callEmergency() {
...
+                                break;
+                            }
+                        }
+                    } else {
+                        throw new InvalidCaseException("The direction cannot be NONE while moving");
                     }
+
+                    System.out.println("Arrived to " + currentFloor + " floor.");
+                    stopElevator();
+                } catch (NoDestinationException e) {
```
[the remark this answers is not in the document]
Comment (bananapizzza) [reply src/main/java/Elevator.java]: Changed.

File: src/main/java/Elevator.java [hunk 11]
```diff
@@ -57,53 +64,59 @@ boolean callEmergency() {
...
+                    } else if (destinationFloor < currentFloor) {
+                        setDirection(Direction.GOING_DOWN);
                     }
-                    currentFloor--;
-                    System.out.println("Current floor is " + currentFloor);
-                    if (destinationFloors[getIndex(currentFloor)]) {
-                        break;
+
+                    if (this.direction.equals(Direction.GOING_UP)) {
+                        while (currentFloor < destinationFloor) {
```
[the remark this answers is not in the document]
Comment (bananapizzza) [reply src/main/java/Elevator.java]: Changed.

**Pull request merged**